Post RA software pipeliner

martien-de-jong opened the pull request:
First workable version of the post-RegAlloc pipeliner.
It creates a DAG that simulates an unrolled loop, and it check whether all latencies can be met by scheduling every copy in the same cycle modulo II.
Since the DAG considers the physical register deps, there's no register renaming to be done, and it uses negative latencies in the same way as the regular post-scheduler.
The results look promising for underfull loops, typically load-operate-store, where the load latency should span quite a few loop iterations to get the resources saturated.

There are still some rough edges, and more tests need to be added, especially to cover the cases where SWP fails.

Commit:
[AIE] Unrelated tidying

Copyright notice, namespace markings, variable casing, ...

Commit:
[AIE] Bring in the PostPipeliner

This is quite an elaborate change, since it is interwoven with the
fixed-point loop in interblock scheduling,

The general approach is to first run loop-aware scheduling, then try to
pipeline selected loops.

When loop aware has converged, each fixed point iteration for the loop block
will increase II until a modulo schedule can be found or it fails.
When we find a modulo schedule with more than one stage, we push out bundled
regions into the prologue block and the epilogue block, update the (only)
region of the loop. The prologue and epilogue regions are copied without
further scheduling when commiting the block schedules.

The pipeliner checks min itercount to guarantee that the LC can be corrected
while staying positive.

There is an unconnected name change, from CurrentBlock to CurrentBlockState
to avoid confusion

We force a fallthrough block for loopend. This avoids missing opportunities
because of 'bad' block ordering

Commit:
[AIE] Reorganize epilogue handling

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -81,10 +81,19 @@ void emitBundlesInScoreboardDelta(
+    // if we have only one, it must be the loop
+    return *MBB.predecessors().begin();
+  }
+  // Otherwise, the loop is the fallthrough predecessor by construction
```
I'm not sure I understand, are we looking for a predecessor of the loop, or for the block which the loop is the predecessor of? I assume the latter given the comment you wrote. Maybe clarify the function name?

Also, how can we guarantee the loop is the fall-through pred, can't we have a branch if MachineBlockPlacement does a bad job?

martien-de-jong replied:
We don't classify something as a loop if the epilogue is not a dedicated exit, unless that exit is fallthrough from the loop. And that classification runs in the same pass. We are looking for the loop, which is a predecessor of the epilogue, and if there are more, it can only be the fallthrough one

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -422,39 +431,79 @@ void InterBlockScheduling::enterRegion(MachineBasicBlock *BB,
+// Create a block, insert it before Succ, and route the control flow edge
+// between Pred and Succ through it.
+// Since we don't add any control flow instructions, the edge should be a
+// fallthrough edge; it will be replaced with two fallthrough edges and a block
```
Where do we verify that the loop falls through to its exit block?

martien-de-jong replied:
In BlockState::classify(), using the layout numbering. If we still have a jump on a fallthrough, we should get rid of that first before thinking about post-pipelining.

Commit:
[AIE] Add a generic 'tripcount update' operand to LoopStart

We use this universally for tripcount update, which allows us
to pipeline loops in both pre and postpipeliner

andcarminati commented on llvm/lib/Target/AIE/AIE2InstrInfo.cpp:
```diff
@@ -1174,12 +1174,21 @@ bool AIE2InstrInfo::isHardwareLoopEnd(unsigned Opcode) const {
+
+void AIE2InstrInfo::adjustTripCount(MachineInstr &MI, int Adjustment) const {
+  assert(MI.getOpcode() == AIE2::ADD_NC);
+  auto &Imm = MI.getOperand(2);
```
Should we assert also for the lc register? Or even more general:
`assert(isZOLTrip....`

andcarminati commented on llvm/lib/Target/AIE/AIE2TargetTransformInfo.h:
```diff
@@ -54,10 +54,10 @@ class AIE2TTIImpl : public BasicTTIImplBase<AIE2TTIImpl> {
-    UP.AllowExpensiveTripCount = true;
     UP.Threshold = 200;
     BaseT::getUnrollingPreferences(L, SE, UP, ORE);
+    UP.Partial = UP.Runtime = false;
```
This is nice! Now our decisions are not changed by BaseT implementation! Maybe those small changes can be merged in small PRs.

andcarminati commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -79,10 +86,19 @@ void emitBundlesInScoreboardDelta(
-  assert(MBB.pred_size() == 1 && "MBB contains more than 1 predecessor");
-  MachineBasicBlock *SinglePredMBB = *MBB.predecessors().begin();
-  return SinglePredMBB;
+MachineBasicBlock *getLoopPredecessor(const MachineBasicBlock &MBB) {
```
nit: good candidate for LoopUtils.

martien-de-jong replied:
I'm hesitant, since it only applies to epilogues that have been selected by the classification of loop-aware candidates here.

andcarminati commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -79,10 +86,19 @@ void emitBundlesInScoreboardDelta(
+  // Otherwise, the loop is the fallthrough predecessor by construction
+  int FallThroughNumber = MBB.getNumber() - 1;
+  for (auto *Pred : MBB.predecessors()) {
+    if (Pred->getNumber() == FallThroughNumber) {
```
Can we use `isLayoutSuccessor` instead of the comparison of getNumbers?

andcarminati commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -297,9 +297,12 @@ void AIEBaseHardwareLoops::expandLoopStart(LowOverheadLoop &LoLoop) {
+
+  // We use ADD_NC, which allows PostPipeliner to tweak it by modifying the
+  // immediate value.
+  BuildMI(*MBB, Start, Start->getDebugLoc(), TII->get(AIE2::ADD_NC), AIE2::LC)
```
We merged e33bbe18b8ed6b608cf77a9c5677d9819ae58563 as a bug fix for the trip count adjustment. As this change also overcomes the same problem, we can replace later (revert the first fix, for example).

andcarminati commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,278 @@
+  // low anyway.
+  // FIXME: Strictly speaking, this is AIE2 specific and should get some
+  // TII support. On the other hand, it will probably remain compatible.
+  if (TripCountDef->getNumOperands() != 3 ||
```
Maybe something like:
```
TII->isNonPipelinedZOLTripCountDef(MI)
```
We can vanish some code here...

martien-de-jong replied:
Done. Included 'Pristine' parameter in isZOLTripCountDef

andcarminati commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -165,33 +170,109 @@ void InterBlockScheduling::leaveFunction() {
+                      << CurrentBlockState->FixPoint.NumIters
+                      << " II=" << CurrentBlockState->FixPoint.II);
+}
+namespace {
```
As this file is growing, maybe we can move this class to AIEPostPipeliner* files.

martien-de-jong replied:
Perhaps not the best criterion to do this. It implements an interface to the pipeliner that is dedicated to the interblock data structures.  All that knowledge would need to be exported to the postpipeliner, which would clutter it.

andcarminati commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -150,33 +170,109 @@ void InterBlockScheduling::leaveFunction() {
+  bool InLoop = false;
+
+  void startPrologue() override {
+    // Nothing at this time, but let's keep the override around
```
Do we have some future plan for this method?

martien-de-jong replied:
Only vaguely. Currently we insert an opaque region which is not seen by the normal scheduler. In future we may need to do some boundary signaling.

andcarminati commented:
Hi @martien-de-jong, some commits from this PR were merged:

https://github.com/Xilinx/llvm-aie/pull/152
https://github.com/Xilinx/llvm-aie/pull/162

Commit:
[AIE] Several fixes and cleanups in ResourceScoreboard

-advance() cleared the wrong cycle
-cleaner interface for reset
-dumper printer required instead of reserved

Commit:
[AIE] Add postpipeliner tests

unittest with complicated binary operator
baseline test for example broken by unnecessary WAW
baseline test for ptradd/load combine with PHI node user
add vmov example

gbossu commented on llvm/include/llvm/CodeGen/ResourceScoreboard.h:
```diff
@@ -63,17 +63,21 @@ template <typename RC> class ResourceScoreboard {
-      Pow2 += Pow2;
-    }
-    if (Cycles.empty()) {
+  void reset(int D = 0) {
```
Nit: How about just introducing a `clear()` method?

martien-de-jong replied:
Yes, sure. Probably an unsupported hunch of many call sites to fix.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -4,14 +4,19 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
-// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+// (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
I don't think this is as old as 2023

gbossu replied:
Did you fix this in a future commit? The diff is still introduced in 46a9ac2bde005b976eee70c59f5f8a98acd0e5db

gbossu commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -179,13 +181,20 @@ class AIEHazardRecognizer : public ScheduleHazardRecognizer {
   unsigned computeScoreboardDepth() const;
 
+  bool checkConflict(const ResourceScoreboard<FuncUnitWrapper> &Scoreboard,
+                     const MCInstrDesc &Desc, const int DeltaCycles) const;
```
Do we need even more of those wrappers? This one seems somewhat error-prone as it ignores register-based itineraries and memory banks.

martien-de-jong replied:
Agreed. This is a level zero resolution of a conflict between Krishnam's PR and mine. I was hesitant to bring in MachineInstr internals here, but I would really like to have a concise conflict checker with an external scoreboard.

martien-de-jong replied:
In fact, it is dead after that rebase.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.h:
```diff
@@ -314,10 +315,17 @@ class InterBlockScheduling {
+  /// Return the number of nops that must be inserted before the epilogue
+  /// of the loop represented by this block.
+  int getSafetyMargin(MachineBasicBlock *Loop,
+                      MachineBasicBlock *Epilogue) const;
```
Why change references to pointers?

martien-de-jong replied:
I don't understand 'change'. This is a new interface. I can make them references. It could be that my first version could have a nullptr Epilogue or Loop to overload the loop and normal interblock case.

gbossu replied:
Correct, `change` does not make sense :D

Still, if you expect the pointers to be non-null, I think a reference is better

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -207,36 +215,111 @@ void InterBlockScheduling::leaveFunction() {
+      if (P == LoopBlock) {
+        continue;
+      }
+      Prologue = &InterBlock.getBlockState(P);
```
Checking: We never re-allocate the vector of `BlockStates`?

martien-de-jong replied:
No, only on first creation. That's part and parcel of the function-level design of interblock

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -381,12 +464,39 @@ bool InterBlockScheduling::updateFixPoint(BlockState &BS) {
+      // Fall back to the loop schedule. Note that we can only have II != 0
+      // after the loop schedule has stabilized.
+      BS.setPipeliningFailed();
+      return true;
```
Doesn't `true` mean that we have converged and do not need to iterate on `MBB` again? In that case don't we want to return `false` to try the "normal" loop scheduling?

gbossu replied:
Hmm looking around, it seems the meaning of the return value is quite overloaded and tied to `isPipelined()` / `isScheduled()`. Maybe `updateFixPoint` could return a `enum State { Failed, TryLoopScheduling, TryModuloScheduling, Done }`?

martien-de-jong replied:
Yeah, makes sense. And that enumeration should replace the different flags in blockstate. I'm still hinging on running the scheduling algorithm as one big fixed-point iteration of all blocks, and then that state determines what to do on enterBlock etc.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -381,12 +465,39 @@ bool InterBlockScheduling::updateFixPoint(BlockState &BS) {
+  // But first try SWP
+  if (BS.getRegions().size() == 1 && BS.getPostSWP().accept(*BS.TheBlock)) {
+    II = 1;
+    return false;
```
Oh so we are first trying the normal "loop iterative scheduling", and once we found a decent schedule, we try to chop it with different II. Do we ever change the "linear schedule" once we reach the `isScheduled()` state? Or do we jus try to take it and "modulo-place" it?

martien-de-jong replied:
Once isScheduled is reached, the regular schedule is fixed, it can only be undone by a succesful postpipelining, which rebuilds the MachineInstr order from scratch. Currently we need one thing from the linear schedule, which is its length, since that determines the number of copies when creating the unrolled DAG. I've been doubting that, since it holds the assumption that this is on par with what the postpipeliner will find. 
The cleanest way would be to generate the nromal single copy DAG, then run scheduleFirstIteration, then use the length from there as stagecount to add the necessary copies to the DAG, then run scheduleOtherIterations.
At this point that would be quite invasive.

gbossu commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -595,9 +604,12 @@ void AIEPostRASchedStrategy::enterRegion(MachineBasicBlock *BB,
   auto &BS = InterBlock.getBlockState(CurMBB);
+  if (BS.FixPoint.II) {
+    return;
+  }
```
Checking: We have an early exit, because when we try to modulo schedule, we do not ever amend the "linear" schedule that is taken as base.

I think somethink like below would really be clearer:
```
if (BS.getState() == TryModuloScheduling)
  return;
```

I don't even think we need a new attribute in BS, it can maybe just be implemented like below.
```
ScheduleState BlocState::getState() {
  if (isPipelined() || (isScheduled() && FixPoint.II > MaxII)
    return Done;
  if (FixPoint.II)
    return TryModuloScheduling;
  if (!isScheduled() && FixPoint.Iters < MaxIters)
    return TryLinearSchedule;
  return Fail;
}
```

This is just making it more explicit that we do have a state machine.  (Note my understanding of that state machine is still quite partial, so my implementation is probably wrong 😄)

martien-de-jong replied:
That state machine should be absolutely clear, since it is interpreted by a disparate set of call backs, and potentially represents the state of every basic block in the function.
I think that IsScheduled, IsPipelined, etc could all be represented by a single enumeration per block, which tell the different scheduler call backs tell what to do.  My gut feeling is that 'BuildRegionStructure' could be added as an initial non-scheduling state, which would probably tidy up some of the awkward locations of doing region management.
enterRegion would just administrate the regions, and enterBasicBlock could decide to iterate over the regions by itself.

gbossu commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -1151,11 +1163,41 @@ void llvm::AIEPostRASchedStrategy::buildGraph(ScheduleDAGMI &DAG, AAResults *AA,
+
+void AIEScheduleDAGMI::schedule() {
+  BlockState &BS = getSchedImpl()->getInterBlock().getBlockState(getBB());
+  if (BS.Kind == BlockType::Loop && BS.FixPoint.II) {
```
Again, here I think `BS.getState() == TryModuleScheduling` would be clearer

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,320 @@
+// all iterations separately, the resources that wrap around accumulate in the
+// overflow area, causing conflicts when inserting future iterations
+int PostPipeliner::fit(MachineInstr *MI, int Earliest, int NTries, int II) {
+  for (int C = Earliest; C < Earliest + NTries; C++) {
```
So we are delaying an instruction until it has no resource conflict? Can we guarantee it is safe given that the linuar schedule has already been critically timed?

martien-de-jong replied:
The intention is not to be dependent on the linear schedule. scheduleFirstIteration is in fact generating a new linear schedule, that is meeting resource and latency. scheduleOtherIterations calls fit() with NTries = 1, which means that it can not delay instructions.
As remarked elsewhere, the linear schedule's length determines the upperbound on the number of stages and the number of copies in the DAG.  I guess we could eliminate any mismatch there by not accepting a schedule with more stages, or rework PostPipeliner to use scheduleFirstIteration for the upperbound of NStages.

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,320 @@
+  }
+  LLVM_DEBUG(dbgs() << "    Fail\n");
+
+  return -1;
```
`std::optional` :)

-1 could be interpreted as a relative cycle

martien-de-jong replied:
And how would you interpret -1 relative to Earliest? :-D

gbossu replied:
Oh you can't (or shouldn't), but an optional value is better than using sentinels I'd say

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,320 @@
+  return -1;
+}
+
+bool PostPipeliner::scheduleFirstIteration() {
```
I'm still a bit confused as to what "first iteration" means here. I assume it's not the first iteration of the non-pipelined loop (which would be e.g. S0+S1+S2), but rather the first iteration of the steady state? Which would be e.g. S2(iter=0) + S1(iter=1) + S0(iter=2)

martien-de-jong replied:
It is actually the first iteration of the non-pipelined loop, and it will probably span all stages of the schedule, though not have instructions in all stages. If we take a simple load+store loop, the load will be in stage 0, the store in stage 7.  With an II of 1, the second iteration will be in stage 1 and stage 8, the third in stage 2 and stage 9, etc
Each next iteration will be shifted by exactly II by the constraints we apply in scheduleOtherIterations.

Commit:
[AIE] Reject PostPipeliner candidates in PrePipeliner

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.h:
```diff
@@ -4,7 +4,7 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
-// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+// (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
Same: please leave as 2024

gbossu commented on llvm/lib/Target/AIE/AIEHazardRecognizer.h:
```diff
@@ -186,6 +186,8 @@ class AIEHazardRecognizer : public ScheduleHazardRecognizer {
 
 protected:
+  ScheduleHazardRecognizer::HazardType getHazardType(const MCInstrDesc &Desc,
+                                                     int DeltaCycles);
```
I don't see that new wrapper being defined. Do we need it?

gbossu commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -545,15 +545,12 @@ void AIEPostRASchedStrategy::enterMBB(MachineBasicBlock *MBB) {
-    TII->insertNoop(*BB, It);
-  }
+  // Safety margin, swp epilogue
+  InterBlock.emitInterBlockTop(BS);
```
We emit epilogue-specific code, shouldn't that be in `emitInterBlockBottom`?

martien-de-jong replied:
The epilogue is pushed into the top of the exit block.

gbossu replied:
I get that, but reading the function name really sounds like "we emit the top of the interblock schedule". I think emitInterblockPrologue and emitInterblockEpilogue is clearer. But nvm, I won't block for that.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.h:
```diff
@@ -314,10 +315,17 @@ class InterBlockScheduling {
-  int getNumEntryNops(const BlockState &BS) const;
+  /// Return the number of nops that must be inserted before the epilogue
+  /// of the loop represented by this block.
+  int getSafetyMargin(MachineBasicBlock *Loop,
```
Nit: safety-margin is very generic, and up to this day, we mostly used it to represent the distance to `ExitSU`. Maybe rename to `getEpilogueSafetyMargin`?

gbossu commented on llvm/test/CodeGen/AIE/aie2/schedule/postpipeliner/bitwisexor.mir:
```diff
@@ -0,0 +1,163 @@
+    successors: %bb.2(0x7c000000), %bb.3(0x04000000)
+    liveins: $p0, $p1, $p2
+
+    $wh0 = VLD_idx_imm_3x32_pseudo $p1, 32 :: (load (<32 x s8>) from %ir.in2.addr.014 + 32, !tbaa !2, addrspace 6)
```
Nit: I'd prefer to use "fake" MMOs like `load from stack - 32` if it means we can get rid of the IR and debloat the file :)

martien-de-jong replied:
Ooohh, clever!

gbossu commented on llvm/test/CodeGen/AIE/aie2/schedule/postpipeliner/load-add-store-renamed.prepiplined.mir:
```diff
@@ -18,21 +18,21 @@
   ; CHECK-NEXT:    nop // Delay Slot 1
   ; CHECK-NEXT:  // %bb.1: // %for.body.preheader
-  ; CHECK-NEXT:    add.nc lc, r0, #-3
+  ; CHECK-NEXT:    add.nc lc, r0, #-1
```
What happened here? I don't see any code change that could trigger this in this commit

martien-de-jong replied:
This is a refactor glitch. I rerun update_tests on all commits and will push shortly.

gbossu commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -173,6 +173,54 @@ char AIEBaseHardwareLoops::ID = 0;
+namespace {
+// Everything becomes much simpler if ZOLs exit to their layout successor
+// We can achieve that by splitting a non-fallthrough exit jump off into
+// a new fallthrough block.
```
This isn't a requirement somewhere in our block placement?

martien-de-jong replied:
There's no 'our block placement'. The pass that's called like that only sets block alignment. The standard block placement is complicated, and choses to let a guard fall through to the exit block. Ultimately, two loops may have the same exit block and one of the two will lose.
Note that the jump would start a new region, which would act like a basic block boundary anyway.

gbossu commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -173,6 +173,54 @@ char AIEBaseHardwareLoops::ID = 0;
+  MachineBasicBlock *TBB, *FBB;
+  const bool AllowModify = false;
+  TII->analyzeBranch(MBB, TBB, FBB, Cond, AllowModify);
+  if (!FBB) {
```
I don't remember how `analyzeBranch` works (who does?), but shouldn't we only trust `TBB` or `FBB` when the branch was successfully analyzed?

martien-de-jong replied:
Yes, that would be better. I will change this.

gbossu commented on llvm/lib/Target/AIE/AIEHazardRecognizer.cpp:
```diff
@@ -441,6 +442,19 @@ ScheduleHazardRecognizer::HazardType AIEHazardRecognizer::getHazardType(
+      Scoreboard, ItinData, SchedClass,
+      getSlotSet(Desc, *TII->getFormatInterface(), IgnoreUnknownSlotSets),
+      MemoryBanks, TII->getMemoryCycles(SchedClass), DeltaCycles, std::nullopt);
+}
```
I'm probably being pedantic, but do we really need one more wrapper if it's called a single time within the new pipeliner code?

martien-de-jong replied:
We could talk about this. I would vouch for dropping the less generic wrappers. I guess that public interfaces using Descr could be dropped in favour of those using MachineInstr. Descr used to be sufficient, but since we use operandinfo to compute memory banks and iteneraries, MachineInstr is more appropriate.

gbossu replied:
We'll fix interfaces in a follow-up. Closing this discussion

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -210,36 +215,100 @@ void InterBlockScheduling::leaveFunction() {
+                      << " II=" << CurrentBlockState->FixPoint.II << "\n");
+}
+namespace {
+class PipelineExtractor : public PipelineScheduleVisitor {
```
Nit: Maybe add a high-level description?

I think we use this as a helper to insert instructions (or their clones) at the right places (prologue, steady state, epilogue)

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -210,36 +215,100 @@ void InterBlockScheduling::leaveFunction() {
+  BlockState *Prologue = nullptr;
+  BlockState *Epilogue = nullptr;
+  MachineBundle CurrentBundle;
+  std::vector<MachineBundle> TimedRegion;
```
Maybe note that `TimedRegion` can represent the epilogue, steady state, or epilogue depending on how far we are in the modulo scheduling.

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,352 @@
+  // (both starting at zero) to the filter
+  auto ExtractSection =
+      [&](int StageCount,
+          std::function<bool(const NodeInfo &Node, int Stage, int M)> Include) {
```
Super-nit: `Filter` instead of `Include` seems more conventional.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -210,36 +215,100 @@ void InterBlockScheduling::leaveFunction() {
+        InLoop ? MI : Loop.TheBlock->getParent()->CloneMachineInstr(MI);
+    CurrentBundle.add(ToBeEmitted);
+  }
+  void endCycle() override { TimedRegion.emplace_back(CurrentBundle); }
```
Super-nit: `Cycle` was confusing to me, because I interpreted it as a CFG cycle, i.e. the steady state. I'd find the interfaces clearer if we had `appendInBundle` and `emitBundle` instead of `startCycle` + `emit` + `endCycle`.

But well, that's probably personal taste at this point :)

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.h:
```diff
@@ -106,14 +107,25 @@ class InterBlockEdges {
+  Pipelining,
+  PipeliningDone,
+  PipeliningFailed
+};
```
Maybe give more details about the states and the legal transitions? In particular:
- Is `SchedulingNotConverged` an error state, or do we still have a way to schedule conservatively? In this case, is it equivalent to `SchedulingDone`?
- Similar questions for `PipeliningFailed`
- Is `SchedulingDone` equivalent to `Pipelining`, or is `SchedulingDone` a final state?

gbossu commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -549,24 +552,33 @@ void AIEPostRASchedStrategy::commitBlockSchedule(MachineBasicBlock *BB) {
-    AIEHazardRecognizer::applyBundles(Region.Bundles, BS.TheBlock);
   }
+  // swp prologue
+  InterBlock.emitInterBlockBottom(BS);
```
Similar comment here: I'm wondering why this is called `emitInterBlockBottom` if we insert things at the top of the block?

Commit:
[AIE2] subtract prologue cycles from 112 byte reservation

gbossu commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -592,9 +604,12 @@ void AIEPostRASchedStrategy::enterRegion(MachineBasicBlock *BB,
-  materializeMultiOpcodeInstrs();
 
   auto &BS = InterBlock.getBlockState(CurMBB);
+  if (BS.FixPoint.II) {
```
I think I still don't understand that early exit. Is it because when trying pipelining, there is no region conflicts to solve? Could that line be changed to `BS.getStage() == Pipelining`?

martien-de-jong replied:
Yep that's an oversight.

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,352 @@
+  for (int C = Earliest; C < Earliest + NTries; C++) {
+    int Mod = C % II;
+    LLVM_DEBUG(dbgs() << "   at " << C << " (" << Mod << ")\n");
+    if (!HR.checkConflict(Scoreboard, *MI, -Depth + Mod)) {
```
Is that really enough to verify conflicts at `-Depth + Mod + II`, `-Depth + Mod + 2*II`, etc.? Or do we have a special scoreboard here that has a size of `II` and wraps around?

martien-de-jong replied:
Hmm. I think I have explained in a number of places that the +II +2*II etc conflicts are accumulating by scheduling all iterations into steady state.

gbossu replied:
Yes I just read the rest of the code now. Coming back here, I can understand it is not needed. There's just a lot of changes to process at once :)

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -343,54 +412,94 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
-      BS.FixPoint.LatencyMargin++;
+  BS.FixPoint.NumIters++;
+  int &II = BS.FixPoint.II;
+  if (!II) {
```
That's now a big function with lots of `if`s. How about introducing two functions for checking the fixpoint state? E.g.

```
BS.FixPoint.NumIters++;

if (State== Scheduling)
  return checkLoopAwareScheduling();
return checkPipelining()
```

gbossu replied:
(That would also help limit the noise in diff lines)

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -343,54 +412,94 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
+                             << " LM=" << BS.FixPoint.LatencyMargin
+                             << " MIM=" << Res.first->second << "\n");
+      // Iterate on CurMBB
+      return BS.FixPoint.Stage = SchedulingStage::Scheduling;
```
Are you voluntarily re-assigning `BS.FixPoint.Stage`?

gbossu replied:
Shouldn't we already be at `SchedulingStage::Scheduling` when starting the loop-aware scheduling?

martien-de-jong replied:
Oh hell.  C++ should require an explicit operator to use an assignment as a value.
But it's not as ugly as I thought, we return the stage, not some boolean.  All returns are of this form to guarantee that the return value is the same as the stage that we save.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -343,54 +412,94 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
+  // This will only work if the stages are small enough. If we miss
+  // opportunities because of this, we should try to get the Prepipeliner
+  // to accept it.
+  if (++II <= 7) {
```
Make that magic number a CL option?

martien-de-jong replied:
Already did that.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -523,6 +633,29 @@ int InterBlockScheduling::getSafetyMargin(MachineBasicBlock *Loop,
+      if (Move) {
+        BB->remove_instr(MI);
+      }
+      BB->insert(Before, MI);
```
Is there an equivalent `insert_instr`? Or is it on purpose so that we insert inside an existing bundle?

martien-de-jong replied:
I can never get my head around these issues. My assumption is that at this point we have no bundles, and we insert instructions before other instructions, leaving BundledWithPred markers to really create the bundles afterwards.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -763,8 +919,13 @@ void BlockState::classify() {
+                                const AIEHazardRecognizer &HR) {
   BoundaryEdges = std::make_unique<InterBlockEdges>(Context);
+  if (Regions.size() == 1) {
+    // Don't worry, this just constructs a mostly empty container class
```
I like this note for the anxious programmer 😄

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,352 @@
+PostPipeliner::PostPipeliner(const AIEHazardRecognizer &HR, int NInstr)
+    : HR(HR), NInstr(NInstr) {}
+
+bool PostPipeliner::accept(MachineBasicBlock &LoopBlock) {
```
Nit: `canAccept`?

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,352 @@
+    MachineInstr *MI = SU.getInstr();
+    const int Earliest = Info[N].Earliest;
+    // The first iteration should fit entirely within II
+    const int Actual = fit(MI, Earliest, II, II);
```
Shouldn't the third parameter be `II - Earliest` if you want to fit within `II`?

gbossu replied:
Oh, I took that comment too literally, obviously we do not want the whole thing to fit within II, otherwise that would be a single stage. I don't think I understand the comment then. What are we protecting ourselves against here?

martien-de-jong replied:
Ah. This comment has some history. fit() actually tries to fit in Earliest % II. If that wraps around, it effectively moves one stage further. So NTries = II just says "Try it in every position in the modulo schedule". More than that is useless.

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,352 @@
+  Info.resize(NTotalInstrs);
+  LLVM_DEBUG(for (int I = 0; I < NInstr;
+                  I++) { dbgs() << I << " " << *DAG->SUnits[I].getInstr(); });
+  computeLoopCarriedEarliest();
```
super-nit: The MachineScheduler does something similar and calls this `bias`. Maybe we could call this function `biasEarliesCyclesForLCDs`?

martien-de-jong replied:
I already renamed this, because it does more.

gbossu commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/split-zol-jmp.mir:
```diff
@@ -0,0 +1,56 @@
+    PseudoJ_jump_imm  %bb.2
+
+  bb.2 (align 16):
+    PseudoRET implicit $lr
```
Nit: if the exit is already dedicated, I guess we do not need a new fall-through? But I guess that is not important because we later optimize the CFG anyway?

martien-de-jong replied:
Well, the fallthrough also makes the body single-regioned, Programming around that is sordid. Also note that hwloop lowering sits just before postmisched.

gbossu commented on llvm/test/CodeGen/AIE/aie2/schedule/postpipeliner/add-store.mir:
```diff
@@ -0,0 +1,130 @@
+  ; CHECK-NEXT:  .LBB0_2: // %for.body
+  ; CHECK-NEXT:    // =>This Inner Loop Header: Depth=1
+  ; CHECK-NEXT:  .L_LEnd0:
+  ; CHECK-NEXT:    nopb ; nopa ; st r1, [p0], #4; add r1, r1, #1; nopm ; nopv
```
🎉

gbossu replied:
🎉

gbossu commented on llvm/test/CodeGen/AIE/aie2/schedule/postpipeliner/bitwisenot.mir:
```diff
@@ -0,0 +1,123 @@
+  ; CHECK-NEXT:    // =>This Inner Loop Header: Depth=1
+  ; CHECK-NEXT:    vldb wh0, [p0, #32]; nopa ; vst wh1, [p1, #32]; nopx ; vbneg_ltz.s16 x1, r21, x0; nopv
+  ; CHECK-NEXT:  .L_LEnd0:
+  ; CHECK-NEXT:    vldb wl0, [p0], #64; nopa ; vst wl1, [p1], #64; nopxm ; nopv
```
Very nice 🎉 🎉

gbossu commented on llvm/lib/Target/AIE/AIEBasePipelinerLoopInfo.cpp:
```diff
@@ -664,6 +669,8 @@ class ZeroOverheadLoop : public AIEBasePipelinerLoopInfo {
 
   bool canAcceptII(SMSchedule &SMS) override;
+
+  bool shouldUseSchedule(SwingSchedulerDAG &SSD, SMSchedule &SMS) override;
```
We should probably update `canAcceptII` as well. Typically, this will increase the II until we have a low enough stage count. So I think that if we find a schedule with a low II and high stage count, we should immediately refuse it sop the post-pipeliner can pick it; and not increase the II.

martien-de-jong replied:
That's an interface change I think? it should be able to say yes, higherII and stop. That would definitely save some time.

gbossu replied:
I guess we would accept the II if we are confident we can pick up the loop in the post-pipeliner. And for those loops, `shouldUseSchedule` would return `false`.

martien-de-jong replied:
Somehow that makes sense, except for the name of canAcceptII. I will add a fat comment.

gbossu replied:
I mean, something can accept it, just not the pre-pipeliner 😄

gbossu reviewed (commented):
I think I'm at a point where I understand all individual pieces, and now I need to go over the code a final time to make sure I understand how it's all connected. But I couldn't find any blocker so far, great work!

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -343,13 +421,32 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
-bool InterBlockScheduling::updateFixPoint(BlockState &BS) {
+SchedulingStage InterBlockScheduling::updateFixPoint(BlockState &BS) {
+  if (BS.Kind != BlockType::Loop) {
+    return BS.FixPoint.Stage = SchedulingStage::SchedulingDone;
```
I still find it confusing to return an assignment. The underlying value of the enum could change (and that will be an annoying thing to debug). How about 
```
BS.FixPoint.Stage = SchedulingStage::SchedulingDone;
return true;
```

Same for other cases when you assign and return at the same time.

gbossu replied:
Clarified, I thought we were implicitly converting the enum to a `bool`, but the return type changed. All good.

gbossu commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -210,36 +219,105 @@ void InterBlockScheduling::leaveFunction() {
-    BS.FixPoint.NumIters++;
+  auto &BS = *CurrentBlockState;
+  switch (updateFixPoint(BS)) {
+  case SchedulingStage::SchedulingNotConverged:
```
I guess this is just for completeness, but we will have crashed before checking `SchedulingStage::SchedulingNotConverged` here?

martien-de-jong replied:
Correct. Hence that I wrote 'may not be observable.` at the enum definition.

gbossu commented on llvm/lib/Target/AIE/AIEMachineScheduler.cpp:
```diff
@@ -1148,11 +1163,41 @@ void llvm::AIEPostRASchedStrategy::buildGraph(ScheduleDAGMI &DAG, AAResults *AA,
+    // Try to wrap the linear schedule within II.
+    // We virtually unroll the body by the stagecount, computed from rounding
+    // up the length divided by II.
+    NCopies = (BS.getScheduleLength() + II - 1) / II;
```
The "standard" schedule has been achieved using different scheduling techniques. Can we guarantee there will be enough copies if the linear schedule used for pipelining is different?

martien-de-jong replied:
Good point. I could make it safe (for now) by rejecting NStages > NCopies. What do you think?

gbossu replied:
Fine to me, I don't think that case will happen very often

gbossu commented on llvm/lib/Target/AIE/AIEPostPipeliner.cpp:
```diff
@@ -0,0 +1,441 @@
+    int Cycle = -Depth + LocalCycle;
+    LLVM_DEBUG(dbgs() << "  Emit in " << Cycle << "\n");
+    HR.emitInScoreboard(Scoreboard, MI->getDesc(), MemoryBanks, MI->operands(),
+                        MI->getMF()->getRegInfo(), Cycle);
```
I'm not debating for correctness, but I wonder if we should not immediately emit the resources at C+II, C+2II, etc. Otherwise, we will only see conflicts due to resources that "wrap around the II" when scheduling the next iterations. At that point, it's too late and we will try a larger II. Maybe if we could anticipate those resources, we could find a better linear schedule that has less conflicts when scheduling all copies.

martien-de-jong replied:
I agree, but that can't be done with the current HazardRecognizer/Scoreboard, since I would need to push the modulo limit into the scoreboard handling. I will create a follow up to do that.

gbossu replied:
I mean, you know the II and NCopies, you could just have a `for` loop here instead of a single call to `HR.emitInScoreboard`? (Not in this PR, agreed)

martien-de-jong replied:
Ah well, it could be done by having enough depth in the scoreboard. Anyway, given that it doesn't impact correctness, I would push that to a follow up.

gbossu reviewed (dismissed):
I'm happy! This is massive change, but it all makes sense and this is a solid base for future improvements. I want it :)

qcolombet commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -173,6 +173,57 @@ char AIEBaseHardwareLoops::ID = 0;
+  DebugLoc DL;
+  TII->insertBranch(MBB, TBB, nullptr, Cond, DL);
+  Cond.clear();
+  TII->insertBranch(*NewBB, FBB, nullptr, Cond, DL);
```
I think it would be easier to read if you use `TII->insertUnconditionalBranch(NewBB, FBB, DL)`.

qcolombet commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -207,36 +219,105 @@ void InterBlockScheduling::leaveFunction() {
+      if (P == LoopBlock) {
+        continue;
+      }
+      Prologue = &InterBlock.getBlockState(P);
```
Is the expectation that we have at most one non-`LoopBlock` predecessor/successor?
If so, could we assert that `Prologue/Epilogue == nullptr` before the assignment?

martien-de-jong replied:
Yes, we check/create these for Loop blocks. The assert would make sense.

qcolombet commented on llvm/lib/Target/AIE/AIEPostPipeliner.h:
```diff
@@ -0,0 +1,155 @@
+  int NInstr;
+  int NCopies;
+  int II = 1;
+  int NStages = 0;
```
Is there a reason that this is `int` and not `unsigned`?

martien-de-jong replied:
Not really. It is a personal preference for the signed integer datatype. For one, you can subtract 'normal' values without underflow.

qcolombet commented on llvm/lib/Target/AIE/AIEPostPipeliner.h:
```diff
@@ -0,0 +1,155 @@
+  /// Place SU in cycle Cycle; update Earliest of dependent instructions
+  void scheduleNode(SUnit &SU, int Cycle);
+
+  /// Compute the stage in which each instruction runs
```
Nit: To be able to upstream such code, all sentences must end with proper punctuatioin. Flagging this to make upstreaming code in general easier :). https://llvm.org/docs/CodingStandards.html#commenting

qcolombet commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -340,13 +421,32 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
-bool InterBlockScheduling::updateFixPoint(BlockState &BS) {
+SchedulingStage InterBlockScheduling::updateFixPoint(BlockState &BS) {
+  if (BS.Kind != BlockType::Loop) {
+    return BS.FixPoint.Stage = SchedulingStage::SchedulingDone;
```
To add on @gbossu returning an assignment is not a common pattern and I too read this as an `==` statement.

My suggestion would be to use the more idiomatic:
```
BS.FixPoint.Stage = SchedulingStage::SchedulingDone;
return BS.FixPoint.Stage;
```

Going beyond (I haven't checked if what I am about to suggest is correct), you could refactor the code to be:
```
if (...) {
  // update BS.FixPoint.Stage
} else if {
  // update BS.FixPoint.Stage differently
} else {
  ...
}
return BS.FixPoint.Stage;
```

I.e., do the updates like you do, but have a single return statement.

Then you can have methods like `updateScheduling` to either return void, or take a `const &` and return the new scheduling state to be assigned to the BlockState.

qcolombet commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -207,36 +219,105 @@ void InterBlockScheduling::leaveFunction() {
-  BS.setScheduled();
-  CurrentBlock = nullptr;
+  CurrentBlockState = nullptr;
   return true;
```
The `return true` for me is confusing because I thought it meant success (and false failure) but we return `true` when the scheduling stage is `PipeliningFailed`. Therefore, from a pure API standpoint it is unclear to me what the users of this API are supposed to do with the result.

I would at least update the comment on leaveBlock (in the header file) to clarify this.

martien-de-jong replied:
Right. It actually means 'fixpoint loop for this block is ready'. Pipelining is the last thing we try, and if we fail, we fall back to the schedule prepared earlier.

qcolombet commented on llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp:
```diff
@@ -340,13 +421,32 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
+
+SchedulingStage InterBlockScheduling::updateScheduling(BlockState &BS) {
+  if (BS.FixPoint.NumIters >
+      MaxExpensiveIterations + 2 * HR->getConflictHorizon()) {
```
I feel it would be useful to keep the previous comment around (assuming, this is still accurate):
```
    // We will first try to increase the latency margin for one instruction at
    // a time, before increasing that margin for all instructions at once.
    // If we are very unlucky, we may step both the latency margin and
    // the resource margin to the max. Any more indicates failure to converge,
    // and we abort to prevent an infinite loop.
```

qcolombet reviewed (commented):
Late to the party and LGTM anyway :).

Just a few comments.
